#########################################################################################################################

# Calculation of true 3D-SFS

nPop1 <- 40  # Number of chromosomes in Pop 1
nPop2 <- 40  # Number of chromosomes in Pop 2
nPop3 <- 40  # Number of chromosomes in Pop 3

readms.output <- "" # Filepath to readms.output.R (download from github.com/dousby/PBS_Project/) 
msoutput <- ""  # Filepath to msoutput.txt, generated by msms

## Reading msms output

source(readms.output)
msmsoutput<- read.ms.output(file=msoutput)

## Separate into different populations

p1.d <- unlist((sapply(msmsoutput$gam,function(x) colSums(x[seq(1,nPop1),]))))
p2.d <- unlist((sapply(msmsoutput$gam,function(x) colSums(x[seq(nPop1+1,nPop1+nPop2),]))))
p3.d <- unlist((sapply(msmsoutput$gam,function(x) colSums(x[seq(nPop1+nPop2+1,nPop1+nPop2+nPop3),]))))

## Calculate pairwise 2D-SFS

pop1.pop2.sfs.2d <- t(sapply(0:nPop1,function(x) table(factor(p2.d[p1.d==x],levels=0:nPop2))))
pop1.pop3.sfs.2d <- t(sapply(0:nPop1,function(x) table(factor(p3.d[p1.d==x],levels=0:nPop3))))
pop2.pop3.sfs.2d <- t(sapply(0:nPop2,function(x) table(factor(p3.d[p2.d==x],levels=0:nPop3))))

## Calculate Real 3D-SFS

Real.3D.SFS <- array(0, dim=c(nPop1+1,nPop2+1,nPop3+1))

for(i in seq(0,nPop3)){
  for(j in seq(0,nPop2)){
    for(k in seq(0, nPop1)){
      p1 <- p1.d[p3.d==i]
      p2 <- p2.d[p3.d==i]
      Real.3D.SFS[i+1,j+1,k+1] <- sum(p1[p2==j]==k)
    }
  }
}

#########################################################################################################################

## RMSD Function

RMSE = function(m, o){
  sqrt(mean((m - o)^2))
}

#########################################################################################################################

## Import ANGSD estimated 3D-SFS into R and calculate RMSD

Est.Real.bias.RMSD <- array(0, dim=c(nPop1+1,nPop2+1,nPop3+1))
Est.Real.bias.summed.RMSD <- numeric(0)

for(i in 5){  ## Where i equal the read depth/s used to calculate GLs
  
  Est.3D.SFS <- paste(“$OUTPUTFOLDER“,i,”/Pop1.Pop2.Pop3.ml", sep="") # Change output folder to own directory
  Est.3D.SFS <- scan(Est.3D.SFS)
  Est.3D.SFS <- array(Est.3D.SFS, dim=c(nPop1+1,nPop2+1,nPop3+1))
  
  Est.Real.bias.RMSD <- RMSE(Est.3D.SFS, Real.3D.SFS)

  Est.Real.bias.summed.RMSD <- c(Est.Real.bias.summed.RMSD,Est.Real.bias.RMSD)
  
}

#########################################################################################################################

## Calculate estimated 3D-SFS for called genotype methods and RMSE calculation

Called.Geno.3D.SFS <- array(0, dim=c(nPop1+1,nPop2+1,nPop3+1))
Called.Geno.RMSD <- array(0, dim=c(nPop1+1,nPop2+1,nPop3+1))
Called.Geno.summed.RMSD <- numeric(0)


  for(i in 5){  # Where i equals the read depth/s used in the GL computation
    print(i)

    ## YRI allele freq by site
    ## Use geno.txt for HWE prior estimation, or geno.uni.txt for uniform distribution prior estimation

    YRI.Freq <- paste("~/Project/Output/RD_Reps_20ind_1Mb/",i,"/YRI.geno.txt", sep="") 
    YRI.Freq <- scan(YRI.Freq)
    YRI.Freq <- matrix(YRI.Freq, nrow = (nPop1/2))
    YRI.Freq <- colSums(YRI.Freq)  
  
    ## CEU allele freq by site
  
    CEU.Freq <- paste("~/Project/Output/RD_Reps_20ind_1Mb/",i,"/CEU.geno.txt",sep="")
    CEU.Freq <- scan(CEU.Freq)
    CEU.Freq <- matrix(CEU.Freq, nrow = (nPop2/2))
    CEU.Freq <- colSums(CEU.Freq)  

    ## CHB allele freq by site

    CHB.Freq <- paste("~/Project/Output/RD_Reps_20ind_1Mb/",i,"/CHB.geno.txt",sep="")
    CHB.Freq <- scan(CHB.Freq)
    CHB.Freq <- matrix(CHB.Freq, nrow = (nPop3/2))
    CHB.Freq <- colSums(CHB.Freq)  

    for(p in seq(0,nPop3)){
      for(j in seq(0,nPop2)){
        for(k in seq(0, nPop1)){
          p1 <- YRI.Freq[CHB.Freq==p]
          p2 <- CEU.Freq[CHB.Freq==p]
          Called.Geno.3D.SFS[p+1,j+1,k+1] <- sum(p1[p2==j]==k, na.rm=TRUE)
        }
      }
    }

## Calculate RSMD of Est - Real

    Called.Geno.RMSD <- RMSE(Called.Geno.3D.SFS,Real.3D.SFS)
    Called.Geno.summed.RMSD <- c(Called.Geno.summed.RMSD, Called.Geno.RMSD)
   
}

#########################################################################################################################

## Importing ANGSD estimated 3D-SFS and calculating RMSD

Est.Real.bias.RMSD <- array(0, dim=c(nPop1+1,nPop2+1,nPop3+1))
Est.Real.bias.summed.RMSD <- numeric(0)

for(i in 5){  ## i equals the read depth used in the GL calculation
  
  Est.3D.SFS <- paste(“OUTPUTFOLDER”,i,”/YRI.CEU.CHB.ml", sep="") ## Change OUTPUTFOLDDER to 
  Est.3D.SFS <- scan(Est.3D.SFS)
  Est.3D.SFS <- array(Est.3D.SFS, dim=c(nPop1+1,nPop2+1,nPop3+1))

  
  Est.Real.bias.RMSD <- RMSE(Est.3D.SFS, Real.3D.SFS)

  Est.Real.bias.summed.RMSD <- c(Est.Real.bias.summed.RMSD,Est.Real.bias.RMSD)
  
}
